include/deal.II/lac/dynamic_sparsity_pattern.h

```diff
@@ -1040,14 +1040,25 @@ DynamicSparsityPattern::begin (const size_type r) const
   // pattern
   //
   // note: row_length(row) returns zero if the row is not locally stored
-  //
-  // TODO: this is way too slow when used in parallel, so do not use it on
-  // non-owned rows
   size_type row = r;
-  while ((row<n_rows())
-         &&
-         (row_length(row)==0))
-    ++row;
+
+  size_type first_own_row = (rowset.size() > 0) ?
+                            rowset.nth_index_in_set(0)
```

Comment:
Shouldn't the check be for `rowset.n_elements() > 0` at this place? I can have an IndexSet whose `size()` is non-zero but still does not have indices and for which `nth_index_in_set` will likely throw.